gax/src/main/java/com/google/api/gax/tracing/TracedBatchingContextCallable.java

```diff
@@ -0,0 +1,123 @@
+import com.google.common.util.concurrent.MoreExecutors;
+
+/**
+ * This callable wraps a batching callable chain in an {@link ApiTracer} and annotates {@link
+ * BatchingCallContext} batching context data.
+ *
+ * <p>For internal use only.
+ */
+@InternalApi("For internal use by google-cloud-java clients only")
+public class TracedBatchingContextCallable<RequestT, ResponseT>
+    extends UnaryCallable<RequestT, ResponseT> {
+
+  private final ApiTracerFactory tracerFactory;
+  private ApiCallContext baseCallContext;
+  private final SpanName spanName;
+  private final UnaryCallable<RequestT, ResponseT> innerCallable;
+
+  public TracedBatchingContextCallable(
+      UnaryCallable<RequestT, ResponseT> innerCallable,
+      ApiCallContext callContext,
+      ApiTracerFactory tracerFactory,
+      SpanName spanName) {
+    this.baseCallContext = Preconditions.checkNotNull(callContext);
+    this.tracerFactory = Preconditions.checkNotNull(tracerFactory);
+    this.spanName = Preconditions.checkNotNull(spanName);
+    this.innerCallable = Preconditions.checkNotNull(innerCallable);
+  }
+
+  /**
+   * Creates an {@link ApiTracer} and annotates batching context data. And perform a call
+   * asynchronously.
+   */
+  public ApiFuture<ResponseT> futureCall(
+      RequestT request, BatchingCallContext batchingCallContext) {
+    ApiTracer tracer =
+        tracerFactory.newTracer(baseCallContext.getTracer(), spanName, OperationType.Batching);
+    TraceFinisher<ResponseT> finisher = new TraceFinisher<>(tracer);
+
+    try {
+      tracer.batchRequestThrottled(batchingCallContext.getTotalThrottledTimeMs());
+      tracer.batchRequestSent(
+          batchingCallContext.getElementCount(), batchingCallContext.getByteCount());
+      baseCallContext = baseCallContext.withTracer(tracer);
+      ApiFuture<ResponseT> future = innerCallable.futureCall(request, baseCallContext);
+      ApiFutures.addCallback(future, finisher, MoreExecutors.directExecutor());
+
+      return future;
+    } catch (RuntimeException e) {
+      finisher.onFailure(e);
+      throw e;
+    }
+  }
+
+  @Override
+  public ApiFuture futureCall(RequestT request, ApiCallContext context) {
+    ApiCallContext mergedContext = baseCallContext.merge(context);
+
+    ApiTracer tracer =
+        tracerFactory.newTracer(mergedContext.getTracer(), spanName, OperationType.Batching);
+    TraceFinisher<ResponseT> finisher = new TraceFinisher<>(tracer);
+
+    try {
+      mergedContext = mergedContext.withTracer(tracer);
+      ApiFuture<ResponseT> future = innerCallable.futureCall(request, mergedContext);
+      ApiFutures.addCallback(future, finisher, MoreExecutors.directExecutor());
+
+      return future;
+    } catch (RuntimeException e) {
+      finisher.onFailure(e);
+      throw e;
+    }
+  }
+
+  public UnaryCallable<RequestT, ResponseT> withDefaultCallContext(
+      final ApiCallContext defaultCallContext) {
+    return new UnaryCallable<RequestT, ResponseT>() {
```

Comment:
we need to be a bit more careful here because it will hide your `futureCall(RequestT, BatchingCallContext)` method. I think you need to have this class store an ApiCallContext and have with `withDefaultCallContext` merge the 2 contexts and return a new instance of TracedBatchingContextCallable